### PR TITLE
Implement configurable handling of excess tokens in datasets

### LINKED ISSUE #775: Option to drop vs truncate on rows longer than context length
### ⚠️ Please check that this feature request hasn't been suggested before.

- [X] I searched previous [Ideas in Discussions](https://github.com/OpenAccess-AI-Collective/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [X] I searched previous [Issues](https://github.com/OpenAccess-AI-Collective/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

Currently for most prompt types, rows with tokens longer than the context currently get dropped. Let's make this customizable. 

### ✔️ Solution

Add an option like excess_len with drop vs truncate as options 

### ❓ Alternatives

_No response_

### 📝 Additional Context

_No response_

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this feature has not been requested yet.
- [X] I have provided enough information for the maintainers to understand and evaluate this request.

### PR DESCRIPTION
**Pull request by mhenrichsen**
- Added `excess_token_handling` option to the configuration, allowing users to choose between "drop" and "truncate" for handling tokens exceeding the maximum sequence length.
- Introduced `truncate_or_drop_long_seq` function to manage both single and batched samples based on the selected handling method.
- Updated relevant dataset processing functions to utilize the new handling option, ensuring backward compatibility with existing "drop" behavior.
- Enhanced logging to reflect truncation actions in dataset processing.

This change improves flexibility in managing sequence lengths during training and evaluation.

# Description
Adds `excess_token_handling`. Either drops or truncates longer sequences.

## How has this been tested?
Ran the following config:
```
base_model: NousResearch/Llama-3.2-1B
# Automatically upload checkpoint and final model to HF
# hub_model_id: username/custom_model_name

datasets:
  - path: teknium/GPT4-LLM-Cleaned
    type: alpaca
dataset_prepared_path: last_run_prepared
val_set_size: 0.1
output_dir: ./outputs/lora-out

adapter: lora
lora_model_dir:

sequence_len: 256
sample_packing: true
eval_sample_packing: true
pad_to_sequence_len: true
excess_token_handling: truncate

lora_r: 16
lora_alpha: 32
lora_dropout: 0.05
lora_target_modules:
  - gate_proj
  - down_proj
  - up_proj
  - q_proj
  - v_proj
  - k_proj
  - o_proj

wandb_project:
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

gradient_accumulation_steps: 2
micro_batch_size: 2
num_epochs: 1
optimizer: adamw_8bit
lr_scheduler: cosine
learning_rate: 0.0002

bf16: auto
tf32: false

gradient_checkpointing: true
resume_from_checkpoint:
logging_steps: 1
flash_attention: true

loss_watchdog_threshold: 5.0
loss_watchdog_patience: 3

warmup_steps: 10
evals_per_epoch: 4
saves_per_epoch: 1
weight_decay: 0.0
special_tokens:
  pad_token: "<|end_of_text|>"
  ```
  
  With `excess_token_handling: truncate` it gives:
  ```
  [2025-05-12 12:32:25,508] [DEBUG] [axolotl.calculate_total_num_steps:624] [PID:5957] [RANK:0] total_num_steps: 805
[2025-05-12 12:32:25,556] [DEBUG] [axolotl.calculate_total_num_steps:513] [PID:5957] [RANK:0] total_num_tokens: 7_449_409
[2025-05-12 12:32:25,856] [DEBUG] [axolotl.calculate_total_num_steps:531] [PID:5957] [RANK:0] `total_supervised_tokens: 4_856_047`
```

And with `drop`:
```
[RANK:0] total_num_steps: 593
[2025-05-12 12:13:11,503] [DEBUG] [axolotl.calculate_total_num_steps:513] [PID:691] [RANK:0] total_num_tokens: 4_549_075
[2025-05-12 12:13:11,747] [DEBUG] [axolotl.calculate_total_num_steps:531] [PID:691] [RANK:0] `total_supervised_tokens: 2_405_977`
```

## Summary by CodeRabbit

* **New Features**
  * Added a configuration option to control how over-length sequences are handled: choose to drop them or truncate to fit.
  * Truncation support applied across pretraining and RL preference datasets, preserving samples by trimming responses when needed.
* **Tests**
  * Added extensive test coverage for sequence overflow handling in both general and RL data flows, including edge cases and batching.
* **Style**
  * Minor log message formatting improvements for optimizer-related output.

### CODE REVIEW

File: docs/config.qmd [hunk 1]
```diff
@@ -332,6 +332,8 @@ dataset_shard_idx:
 # The maximum length of an input to train with, this should typically be less than 2048
 # as most models have a token/context limit of 2048
 sequence_len: 2048
+# How to handle tokens exceeding max sequence length - "drop" (default, removes sample) or "truncate" (cuts off excess tokens)
+excess_token_handling: drop
```
Comment (winglian) [inline docs/config.qmd]: maybe something like `sequence_len_excess` or `sequence_len_overflow` would be obvious

File: src/axolotl/utils/data/pretraining.py [hunk 2]
```diff
@@ -259,6 +259,8 @@ def encode_packed_pretraining(
         # FIXME using attention mask unpad/pad with trainer and packed pretraining is broken atm
         # workaround by using the position id logic for now in trainer
         drop_attention_mask=multipack_attn,
+        # pass through handling mode from config via ds_wrapper function
+        handling=getattr(ds_wrapper, "cfg", {}).get("excess_token_handling", "drop"),
```
Comment (winglian) [inline src/axolotl/utils/data/pretraining.py]: maybe somewhere in constants we have a DEFAULT_{whatever we name_it}_VALUE = "drop" and can reference it throughout

File: src/axolotl/utils/data/rl.py [hunk 3]
```diff
@@ -96,11 +98,51 @@ def drop_long_rl_seq(
...
-        return (len_prompt + len_chosen) <= sequence_len and (
-            len_prompt + len_rejected
-        ) <= sequence_len
+        if handling == "drop":
+            result = (len_prompt + len_chosen) <= sequence_len and (
+                len_prompt + len_rejected
+            ) <= sequence_len
+
+        # truncate
+        else:
```
Comment (winglian) [inline src/axolotl/utils/data/rl.py]: since the original default is drop, my preference is `if truncate; do; else` so the fallback is always the drop functionality.

File: src/axolotl/utils/trainer.py [hunk 4]
```diff
@@ -370,15 +478,31 @@ def drop_no_trainable_tokens(sample):
...
+    drop_long_fn = partial(drop_long_seq, sequence_len=sequence_len)
+
+    # Use filter for drop mode and map for truncate mode
+    if handling == "drop":
+        train_dataset = train_dataset.filter(
+            drop_long_fn,
+            desc="Dropping Long Sequences",
+            load_from_cache_file=False,
+        )
+    else:
```
Comment (winglian) [inline src/axolotl/utils/trainer.py]: Is this needed if the new function handles both?

**Review by winglian — COMMENTED** [review]
Would be good to add a unit test that runs a dataset through and checks the lengths of the outputs and counts (drop should have fewer rows whereas truncate should have the same count as the input dataset)

File: src/axolotl/utils/trainer.py (hunk 4 above)
Comment (mhenrichsen) [reply src/axolotl/utils/trainer.py]: You're right, the separate filter logic isn't needed anymore. I've updated `process_pretraining_datasets_for_packing` to use the same `truncate_or_drop_long_seq` function for both modes. It now returns a boolean when `handling="drop"` (for use with `.filter()`) and the modified sample when `handling="truncate"` (for use with `.map()`), consolidating the logic as you suggested.

File: src/axolotl/utils/data/pretraining.py (hunk 2 above)
Comment (mhenrichsen) [reply src/axolotl/utils/data/pretraining.py]: Where would you want this placed?

Comment (mhenrichsen) [reply src/axolotl/utils/data/pretraining.py]: @winglian we just need this one then we are good to go I think

Comment (NanoCode012) [reply src/axolotl/utils/data/pretraining.py]: @mhenrichsen , let's put it in the adjacent `utils.py`?
